File: 2022-go/day01.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

func topThree(filename string) []int {
	fin, _ := os.Open(filename)
	defer fin.Close()
	sc := bufio.NewScanner(bufio.NewReader(fin))
	sum := 0
	var nums []int
	for sc.Scan() {
		t := sc.Text()
		if len(t) > 0 {
			x, _ := strconv.Atoi(t)
			sum += x
		} else if sum != 0 {
			nums = append(nums, sum)
			sum = 0
		}
	}
	nums = append(nums, sum)
	// https://stackoverflow.com/a/18343326/2990344
	sort.Sort(sort.Reverse(sort.IntSlice(nums)))
	return nums[0:3]
}

func main() {
	t := topThree("/home/xdavidliu/Documents/aoc/input01.txt")
	fmt.Println("part 1 =", t[0])            // 66306
	fmt.Println("part 2 = ", t[0]+t[1]+t[2]) // 195292
}

File: 2020/day01.jl
a = open("/home/xdavidliu/Documents/aoc/input01.txt") do f
    [parse(Int, s) for s in readlines(f)]
end

function partone(a)
    for i in eachindex(a)
        for k = i+1:length(a)
            if a[i] + a[k] == 2020
                return a[i] * a[k]
            end
        end
    end
    error()
end

function parttwo(a)
    for i in eachindex(a)
        for j = i+1:length(a)-1
            for k = j+1:length(a)
                if a[i] + a[j] + a[k] == 2020
                    return a[i] * a[j] * a[k]
                end
            end
        end
    end
end

println("part 1 = ", partone(a))  # 355875
println("part 2 = ", parttwo(a))  # 140379120
